tests/test_validity.py

```diff
@@ -64,93 +65,125 @@ def periodic_disk(n, s):
+                {StorageType.RAM: s, StorageType.DISK: 0}, 1)
 
 
 def mixed(n, s):
     return (MixedCheckpointSchedule(n, s),
             {StorageType.RAM: 0, StorageType.DISK: s}, 1)
 
 
+def single_disk_copy(n, s):
+    cp_schedule = SingleDiskStorageSchedule(move_data=False)
+    return (cp_schedule,
+            {StorageType.RAM: 0, StorageType.DISK: n}, 1)
+
+
+def single_disk_move(n, s):
+    cp_schedule = SingleDiskStorageSchedule(move_data=True)
+    return (cp_schedule,
+            {StorageType.RAM: 0, StorageType.DISK: n}, 1)
+
+
+def single_memory(n, s):
+    cp_schedule = SingleMemoryStorageSchedule()
+    return (cp_schedule,
+            {StorageType.RAM: 0, StorageType.DISK: 0}, n)
+
+
 @pytest.mark.parametrize(
     "schedule",
     [
-     revolve,
-     periodic_disk,
-     disk_revolve,
      h_revolve,
+     disk_revolve,
      multistage,
      twolevel_binomial,
+     periodic_disk,
+     revolve,
      mixed,
+     single_disk_copy,
+     single_disk_move,
+     single_memory
      ]
      )
 @pytest.mark.parametrize("n, S", [
-                                  (5, (2,)),
+                                  (1, (0,)),
+                                  (2, (1,)),
                                   (3, (1, 2)),
+                                  (5, (2,)),
                                   (10, tuple(range(2, 10))),
                                   (100, tuple(range(1, 100))),
                                   (250, tuple(range(25, 250, 25)))
                                   ])
 def test_validity(schedule, n, S):
-    """Test the checkpoint revolvers.
+    """Test validity of checkpoint schedules. Tests that an adjoint calculation
+    can be performed without exceeding storage limits.
 
     Parameters
     ----------
-    schedule : object
-        Revolver schedule.
+    schedule : callable
+        Accepts the number of forward steps and checkpoint units, and returns a
+        schedule.
     n : int
-        Total forward steps.
+        Number of forward steps.
     S : int
-        Snapshots.
+        Number of checkpoint units.
     """
+
     @functools.singledispatch
     def action(cp_action):
         raise TypeError("Unexpected action")
 
     @action.register(Forward)
     def action_forward(cp_action):
         nonlocal model_n
-        # Start at the current location of the forward
-        assert model_n is not None and model_n == cp_action.n0
 
+        # Start at the current location of the forward
+        assert model_n is not None and cp_action.n0 == model_n
         # If the schedule has been finalized, end at or before the end of the
         # forward
         assert cp_schedule.max_n is None or cp_action.n1 <= n
+
         if cp_schedule.max_n is not None:
             # Do not advance further than the current location of the adjoint
             assert cp_action.n1 <= n - model_r
-
         n1 = min(cp_action.n1, n)
 
         model_n = n1
         ics.clear()
-        data.clear()
```

Comment:
Do we assume data in `StorageType.WORK` is replaced by a `Forward` action? i.e. only data recorded during the `Forward` action is retained in `StorageType.WORK` at the end of the `Forward` action? If so `data.clear()` can be retained here.

Reply:
We have `StorageType.WORK` in the `Copy` and `Move` actions as well, e.g., `Move(2, StorageType.RAM, StorageType.WORK)`.

Reply:
The question is whether
```
Forward(0, 1, False, True, StorageType.WORK)
Forward(0, 2, False, True, StorageType.WORK)
EndForward()
Reverse(2, 1, False)
Reverse(1, 0, False)
EndReverse()
````
should be a valid schedule. If the validity test uses `data.clear()` then this implies the above is an _invalid_ schedule.

Reply:
I've updated the test, adding `data.clear()`, assuming that the previous example should be invalid.

Reply:
Ah yes. I do not see a case where we can have 
```
Forward(0, 1, False, True, StorageType.WORK)
Forward(1, 2, False, True, StorageType.WORK)
```
The Forward action like `Forward(n0, n1 False, True, StorageType.WORK)` is set when we are in the reverse computation. So, the next action must be `Reverse`.